feat(Chip): Add chip component

cezarykubel opened the pull request:
A new chip component, wrapped in a auto-collapsing chip list component.  This was originally used for AD-4010, the topics in the details view work.  

Some examples of the different options:
![image](https://user-images.githubusercontent.com/7015464/56316841-58084080-6129-11e9-846e-e6f819c3c5e2.png)
![image](https://user-images.githubusercontent.com/7015464/56316852-60607b80-6129-11e9-927f-969b87d6e68c.png)
![image](https://user-images.githubusercontent.com/7015464/56316861-68b8b680-6129-11e9-8982-ea68f510684f.png)

Some things to note:
- if there are not enough chips to have a multiline list, the list will not be collapsible
- you can use the `size` prop on both individual Chip components and the whole ChipList component.  The latter will apply the size to all Chips within the container.  
- if `activeColor` is not set, the default is 'accent' and solid.  However, if an `activeColor` is set, the Chip has a "outline" style, as per other use-cases.  

Preview [here](https://deploy-preview-740--serene-shannon-8198f2.netlify.com/docs/packages/axiom-components/chip)

lpoulter commented on packages/axiom-components/src/Chip/ChipList.css:
```diff
@@ -0,0 +1,72 @@
+  margin-top: var(--spacing-grid-size--x2);
+  margin-bottom: var(--spacing-grid-size--x2);
+  overflow: hidden;
+  &::after {
```
If we remove the floats and use flex on the list can we also remove the clearfix? 

```
.ax-chiplist {
		display: flex;
    	flex-wrap: wrap;
}
```

cezarykubel replied:
fixed.  not quite sure why i used floats.  old habits? :/

lpoulter commented on packages/axiom-components/src/Chip/ChipList.css:
```diff
@@ -0,0 +1,72 @@
+  align-items: center;
+  justify-content: center;
+  margin-right: var(--spacing-grid-size--x1);
+  float: left;
```
I'm not sure we need to float this if we use flex?

cezarykubel replied:
fixed

larister commented:
Congrats on sneaking in some Celtics players into Axiom 😂

larister commented on packages/axiom-components/src/Chip/ChipList.css:
```diff
@@ -0,0 +1,72 @@
+.ax-chiplist--large {
+  transition-property: max-height, opacity, margin;
+  transition-duration: var(--transition-time-slow);
+  transition-timing-function: var(--transition-function);
```
Can't quite figure out why but it looks to me that on the demo at least the transition is only happening on "collapse", not on "expand"

cezarykubel replied:
its because we use `cubic-bezier`, specifically `cubic-bezier(0.25, 0.46, 0.45, 0.94)` for all transitions.  when the list is large, the transition is smoother on both the collapse and expand

larister replied:
hmm interesting, thanks. Still a bit surprised it only happens on "collapse" not "expand" for the demo but nabd

larister commented on packages/axiom-components/src/Chip/ChipList.js:
```diff
@@ -0,0 +1,86 @@
+      isExpandable,
+    } = this.state;
+
+    const childrenWithSize = React.Children.map(children, child => React.cloneElement(child, { size }));
```
Cool, didn't know you could do that. Part of me wonders if it's worth checking the child is actually a chip but not sure it's worth it

cezarykubel replied:
I don't think i've needed to use it before but its super useful, especially in this use-case when with one line of code, you can set the size of all the Chips with one prop on the container.  As far as checking the child, technically, it should accept any list of components, Chip or not, so I don't know if thats necessary.  I can be easily swayed otherwise if everyone thinks otherwise.

larister replied:
Nah I don't think it's worth it personally

larister commented on packages/axiom-components/src/Chip/ChipList.js:
```diff
@@ -0,0 +1,86 @@
+
+        <div className={ labelClasses } onClick={ this.handleExpand }>
+          { label }
+          { isExpandable && (<Icon inline name={ isExpanded ? 'chevron-up' : 'chevron-down' } spaceLeft="x1" />) }
```
would slightly prefer this was abstracted out into a fn rather than having multiple conditions inline like this but nabd

cezarykubel replied:
hrmm...would that just cause more jumbled code, trying to find the functions relating to each bit of the render method?

larister replied:
yeah maybe you're right - nabd anyway

lpoulter replied:
I've gone back and forth on this but generally prefer the way Cezary has done it here as theres really not much logic.

larister commented on packages/axiom-components/src/Chip/ChipList.js:
```diff
@@ -0,0 +1,86 @@
+      scrollHeight,
+    } = this.chipList.current;
+
+    if (scrollHeight > (clientHeight * 2)) {
```
Could we use `getBoundingClientRect()` instead? Feels like the multiplication by 2 is a little inexact; I might have misunderstood though

cezarykubel replied:
I don't think I've used getBoundingClientRect() before but it seems to get the same values as `clientHeight`.  The reasoning behind using 2x the client height is because we want to make sure that the ChipList is only expandable IF theres more than one line of chips.  The client height represents one line of chips

larister replied:
Huh, it's meant to take into account padding/margin etc AFAIK and it did seem to give a different value to `clientHeight` when I tried it.

> The reasoning behind using 2x the client height is because we want to make sure that the ChipList is only expandable IF theres more than one line of chips. The client height represents one line of chips

Sorry I'm being slow here, but in that case shouldn't it be` >=`?

Don't worry about answering if you don't want, this code works and is fine, it's just for my own curiosity really :)

cezarykubel commented on packages/axiom-components/src/Chip/Chip.js:
```diff
@@ -0,0 +1,101 @@
+    /** Apply active styling */
+    active: PropTypes.bool,
+    /** Colour of the Chip when active */
+    activeColor: PropTypes.oneOf([
```
you'll never convince me 😄

larister replied:
Not sure I'll ever convince myself...

cezarykubel commented on packages/axiom-components/src/Chip/Chip.css:
```diff
@@ -0,0 +1,116 @@
+  margin-right: var(--cmp-inline-spacing);
+  margin-bottom: var(--cmp-inline-spacing);
+  padding: 0;
+  float: left;
```
yeeeup, not quite sure why i used floats initially.  but its fixed now

cezarykubel commented:
made a few adjustments:  removed the colored chips and fixed the transition based on the design review.  also, fixed the disabled state for active chips.
